log_entry.go

```diff
@@ -0,0 +1,242 @@
+func (entry *LogEntry) AsFatal() *LogEntry {
+	return entry.AsLevel(FatalLevel)
+}
+
+// AsPanic clones the entry into a new log entry and sets the level to `panic`
+// Make sure you call this method before calling WithField, WithFields and WithError methods
+func (entry *LogEntry) AsPanic() *LogEntry {
+	return entry.AsLevel(PanicLevel)
+}
+
+// WithField adds a field to the log entry, note that it doesn't log until you call Write.
+func (entry *LogEntry) WithField(key string, value interface{}) *LogEntry {
+	if entry.Level > entry.Logger.level() {
+		return entry
+	}
+	//Do not change this to Fields{key:value}. You will end up getting more allocations
+	fields := make(Fields, 1)
+	fields[key] = value
+	return entry.WithFields(fields)
+}
+
+// WithFields adds a struct of fields to the log entry
+func (entry *LogEntry) WithFields(fields Fields) *LogEntry {
+	if entry.Level > entry.Logger.level() {
+		return entry
+	}
+	data := make(Fields, len(entry.Data)+len(fields))
+	for k, v := range entry.Data {
+		data[k] = v
+	}
+	for k, v := range fields {
+		data[k] = v
+	}
+	return &LogEntry{
+		Data:   data,
+		Level:  entry.Level,
+		Logger: entry.Logger,
+	}
+}
+
+// WithError adds an error as single field to the log entry
+func (entry *LogEntry) WithError(err error) *LogEntry {
+	return entry.WithField(ErrorKey, err)
+}
+
+func (entry *LogEntry) Writef(format string, args ...interface{}) {
+	entry.write(formatted, format, args...)
+}
+
+func (entry *LogEntry) Write(args ...interface{}) {
+	entry.write(unformatted, "", args...)
+}
+
+func (entry *LogEntry) Writeln(args ...interface{}) {
+	entry.write(newLine, "", args...)
+}
+
+func (entry *LogEntry) write(mode formatMode, format string, args ...interface{}) {
+	if entry.Logger.level() >= entry.Level {
+		message := constructMessage(mode, format, args...)
+		entry.log(entry.Level, message)
+	}
+}
+
+func constructMessage(mode formatMode, format string, args ...interface{}) string {
+	switch mode {
+	case formatted:
+		return fmt.Sprintf(format, args...)
+	case unformatted:
+		return fmt.Sprint(args...)
+	case newLine:
+		return sprintlnn(args...)
+	}
+	return fmt.Sprintf(format, args...)
+}
+
+// This function is not declared with a pointer value because otherwise
+// race conditions will occur when using multiple goroutines
+func (entry LogEntry) log(level Level, msg string) {
+	var buffer *bytes.Buffer
+	entry.Time = time.Now()
+	entry.Level = level
+	entry.Message = msg
+
+	entry.Logger.mu.Lock()
+	err := entry.Logger.Hooks.Fire(level, &entry)
+	entry.Logger.mu.Unlock()
+	if err != nil {
+		entry.Logger.mu.Lock()
+		fmt.Fprintf(os.Stderr, "Failed to fire hook: %v\n", err)
+		entry.Logger.mu.Unlock()
+	}
+	buffer = bufferPool.Get().(*bytes.Buffer)
+	buffer.Reset()
+	defer bufferPool.Put(buffer)
+	entry.Buffer = buffer
+	serialized, err := entry.Logger.Formatter.FormatEntry(&entry)
+	entry.Buffer = nil
+	if err != nil {
+		entry.Logger.mu.Lock()
```

Comment:
one global lock , or we can change it to use channel, and use one go routine to do the logging

Reply:
I would leave it as it is for the time being because as discussed, it will change the interface dramatically. Besides, locking is not the costly part I reckon. Formatting is the most expensive function at the moment